### ISSUE TITLE
Motivate Pytorch's forward mode AD APIs with training examples

### ISSUE THREAD

**Issue by redwrasse** [pytorch/pytorch#152877]
### 📚 The doc issue

Pytorch documentation of forward AD (https://pytorch.org/tutorials/intermediate/forward_ad_usage.html) might benefit from training examples for the uninitiated user. The current documentation explains how to compute JVPs, though without context.

Adding training examples may also add community cycles to testing and tweaking the forward AD support (*_jvp operators) Pytorch has invested in: https://github.com/pytorch/pytorch/issues/71117

Textbook context, for example: (Chapter 6 of Deep Learning (Goodfellow, Bengio, Courville)): "_when the number of outputs of the graph is larger than the number of inputs, it is sometimes preferable to use another form of automatic differentiation called forward mode accumulation. Forward mode accumulation has been proposed for obtaining real-time computation of gradients in recurrent networks, for example (William and Zipserr, 1989).  This approach also avoids the need to store the values and gradients for the whole graph, trading off computational efficiency for memory."_

### Suggest a potential alternative/fix

Something that illustrates training, with computation and memory tradeoffs against the equivalent architecture via backprop, would be useful.

Stuck this prompt into Google Gemini: 
```
""Code me up a Pytorch code sample of training an RNN using forward mode AD, following the ideas of the Williams and Zipser paper here:
# https://web.stanford.edu/class/psych209a/ReadingsByDate/02_25/Williams%20Zipser95RecNets.pdf"
```

which returned syntactically incorrect code, but nonetheless illustrates the idea of training with forward AD:

```
# Auto-generated by Google Gemini:
# "Code me up a Pytorch code sample of training an RNN using forward mode AD, following the ideas of the Williams and Zipser paper here:
# https://web.stanford.edu/class/psych209a/ReadingsByDate/02_25/Williams%20Zipser95RecNets.pdf"
import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd.functional import jvp

# --- Hyperparameters ---
input_size = 10
hidden_size = 5
output_size = 1
seq_len = 5
batch_size = 2
learning_rate = 0.01
num_epochs = 1

# --- Define the RNN Model ---
class SimpleRNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(SimpleRNN, self).__init__()
        self.rnn = nn.RNNCell(input_size, hidden_size)
        self.linear = nn.Linear(hidden_size, output_size)

    def forward(self, input_, hidden):
        hidden = self.rnn(input_, hidden)
        output = self.linear(hidden)
        return output, hidden

    def init_hidden(self, batch_size):
        return torch.zeros(batch_size, hidden_size)

# --- Generate Synthetic Data ---
def generate_data(batch_size, seq_len, input_size, output_size):
    inputs = torch.randn(seq_len, batch_size, input_size)
    targets = torch.randn(seq_len, batch_size, output_size)
    return inputs, targets

inputs, targets = generate_data(batch_size, seq_len, input_size, output_size)

# --- Instantiate the Model and Optimizer ---
model = SimpleRNN(input_size, hidden_size, output_size)
optimizer = optim.SGD(model.parameters(), lr=learning_rate)
criterion = nn.MSELoss()

# --- Forward Mode AD Training Loop (Inspired by Williams & Zipser) ---
for epoch in range(num_epochs):
    hidden = model.init_hidden(batch_size)
    total_loss = 0

    for t in range(seq_len):
        input_t = inputs[t]
        target_t = targets[t]

        # --- Define the forward pass function for JVP ---
        def forward_pass(params):
            # Unpack parameters (order matters and should match model.parameters())
            (weight_ih, weight_hh, bias_ih, bias_hh, weight_ho, bias_ho) = params

            # Initialize hidden state (we need to do this inside for independent JVP)
            current_hidden = torch.zeros(batch_size, hidden_size)

            # Perform forward pass up to the current time step
            for i in range(t + 1):
                input_step = inputs[i]
                current_hidden = torch.tanh(torch.matmul(input_step, weight_ih.T) + bias_ih +
                                            torch.matmul(current_hidden, weight_hh.T) + bias_hh)

            output = torch.matmul(current_hidden, weight_ho.T) + bias_ho
            loss = criterion(output, target_t)
            return loss

        # --- Define tangent vectors for each parameter (for a single example in the batch) ---
        tangents = tuple(torch.randn_like(p) for p in model.parameters())

        # --- Compute JVP for each example in the batch ---
        jvp_loss_per_example = []
        for b in range(batch_size):
            def forward_pass_single(params):
                (weight_ih, weight_hh, bias_ih, bias_hh, weight_ho, bias_ho) = params
                current_hidden = torch.zeros(1, hidden_size)
                for i in range(t + 1):
                    input_step = inputs[i][b].unsqueeze(0)
                    current_hidden = torch.tanh(torch.matmul(input_step, weight_ih.T) + bias_ih +
                                                torch.matmul(current_hidden, weight_hh.T) + bias_hh)
                output = torch.matmul(current_hidden, weight_ho.T) + bias_ho
                loss_single = criterion(output, target_t[b].unsqueeze(0))
                return loss_single

            (loss_output_single, jvp_output_single) = jvp(forward_pass_single, tuple(model.parameters()), tangents)
            jvp_loss_per_example.append(jvp_output_single)

        # --- Aggregate JVP over the batch (simple sum for this example) ---
        aggregated_jvp = tuple(torch.stack([jvp[i] for jvp in jvp_loss_per_example]).mean(dim=0) for i in range(len(tangents)))

        # --- "Update" parameters based on the JVP (simplified SGD-like update) ---
        with torch.no_grad():
            for param, grad_v in zip(model.parameters(), aggregated_jvp):
                param.sub_(learning_rate * grad_v)

        # --- Standard forward pass for loss tracking ---
        output, hidden = model(input_t, hidden)
        loss = criterion(output, target_t)
        total_loss += loss.item()

    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {total_loss / seq_len}')

print("Training finished (one epoch).")
```
